### ISSUE TITLE
input[number] automatic recognition of localized decimal character not consistent across browsers

### ISSUE THREAD

**Issue by zzal** [angular/angular.js#10146]
You can try it from the official API docs: [https://docs.angularjs.org/api/ng/input/input[number]](https://docs.angularjs.org/api/ng/input/input%5Bnumber%5D)

I don't know if this is a bug or a lack of parameters adjustments, but I post it anyway. 

When you're in a non-english localized browser (in which the decimal mark is the comma, like in French, German, Finnish, etc), if you enter a valid number (say 12,5 or 12.5), the Angular input[number] directive recognize it like a valid number. 
But as soon as you use an english-only browser, there is no way Angular can treat the same input as valid. If you enter 12,5 in the same input field, it considers it as invalid.
I think this is not consistent, and there should be a way to tell the directive what characters are considered as the decimal separator. Or maybe there is a way to set this but it is not documented.

**Comment by Psysoul** [comment]
@zzal Yes, the keyboard layout change in case of [type=number] is quite important to provide good UX for the user.

This issue is quite important to be fixed, as now with [type=number] we can't enable users to enter either comma or period as decimal separator, as in case of comma the model value is undefined. Angularjs should enable developers to setup how comma separators work or just provide a way to customize all the details.

What I would like to achieve is to enable users to enter comma and period in the input, and handle them based on the currently selected user locale, so for example: 

1. Current user locale decimal separator: period, then:
Thousand separator is comma, remove all of them from input, use period as decimal separator (should be done by js by default).

2. Current user locale decimal separator: comma, then:
Thousand separator is period: remove all of them from input: convert comma to period to be used as decimal separator.

Basically the same thing as mentioned before.

**Comment by konsultaner** [comment]
@zzal The directive does not catch the 1000-separator. A german ```1.000,12``` Should be parsed to ```1000.12``` but instead is parsed to ```1.00012```

**Comment by SAIsalim** [comment]
+1

**Comment by Narretz** [comment]
@konsultaner That happens for browser that don't support input[number], right?

**Comment by konsultaner** [comment]
@Narretz I usually work with chrome, so no this happens in general I guess.

**Comment by Thariq-Nugrohotomo** [comment]
After following the suggestion from @lgalfaso , I have successfully made the `input[number]` in IE11 can accept either Comma or Period as Decimal-Separator.

Roughly the directive will look like:
```javascript
restrict:'A',
require:'?ngModel', // get a hold of NgModelController
link:function ($scope, $element, $attrs, $ngModel) {
    if ($ngModel) {
        $ngModel.$parsers.unshift(function (str) {
            if (str && (typeof str === 'string' || str instanceof String)) {
                // sanitize : replace Comma with Period/Dot
                return str.replace(',', '.');
            } else {
                // not a string, can't process it, give up, will be handled by next parser within pipeline
                return str;
            }
        });
    }
}
```

I think it's better to just force the user to use Period though, e.g. by rejecting Comma within `keypress` or `keydown`.
Browser's behavior for Comma is varies, even worse the behavior of the same browser can be different if the language settings are different.

**Comment by robinwassen** [comment]
@Thariq-Nugrohotomo I did some testing of your solution, it works in IE11 as you say, just want to give a heads up that this does not work as intended in Chrome, Win and Edge, Win.

- Chrome rejects the comma input and therefore it does not reach the angular parser.

- Edge accepts the input but does not send a value to AngularJS if the decimal number contains a comma.

The only solution that I can think of with these facts to work in all desktop browsers is to use input[type=text] instead and build manual parsing and validation of the number.

If mobile is needed I think the best full blown solution would be to fall back to using a input[type=number] if it is a mobile device to get a proper number keyboard when the user enters a number.

**Comment by Thariq-Nugrohotomo** [comment]
@robinwassen I haven't do any testing for Edge,
but as for Chrome, I just tell the client to switch their Chrome language settings to the one that supports comma as decimal-separator (e.g. french, german). They also need to make Chrome being shown in that selected language. This way, Chrome will accept the comma, even without adding custom parser.

Custom parser that I mentioned above is just for IE (tested in IE11).

**Comment by petebacondarwin** [comment]
This is a general problem with browsers that support `input[number]` since the value that AngularJS receives has already been converted by the browser to a number using the browser's locale settings.

**Comment by MatteoSp** [comment]
I'm confused, can someone help me understand what should I do when:
1. My users use a mix of browsers/language settings
2. They just want the decimal separator to ALWAYS be "," (comma)

Why including the proper angular-locale_* is not enough?

**Comment by robinwassen** [comment]
@MatteoSp - Use <input type="text />, build custom validation logic for numbers with comma decimal separator. 

Due to how browsers work it is impossible to achieve with input type="number".

**Comment by zian974** [comment]
Old post, but maybe a solution.
The locale of my web app is fr_FR. I use the attribute lang="en_EN" on input type=number. Firefox display the float with a dot and Chrome display the float with a comma but I can use indifferently the comma and the dot, no pb of validation

**Comment by levindoneto** [comment]
> @MatteoSp - Use <input type="text />, build custom validation logic for numbers with comma decimal separator.
> 
> Due to how browsers work it is impossible to achieve with input type="number".

This isn't a solution since all the key events would be messed-up, needing a considerable overhead to be handled.
